service/src/main/java/bio/terra/pipelines/service/PipelinesService.java

```diff
@@ -51,4 +55,108 @@ public Pipeline updatePipelineWorkspaceId(PipelinesEnum pipelineName, UUID works
+          String.format("Problems with pipelineInputs: %s", String.join("; ", errorMessages)));
+    }
+  }
+
+  private LinkedHashMap<String, Object> castInputsToMap(Object inputs) {
+    try {
+      return objectMapper.convertValue(inputs, LinkedHashMap.class);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("pipelineInputs must be a JSON object");
+    }
+  }
+
+  /**
+   * Validate that all required inputs are present in the inputsMap
+   *
+   * @param inputDefinitions - list of input definitions for a pipeline
+   * @param inputsMap - map of inputs to validate
+   * @return list of error messages for missing required inputs
+   */
+  public List<String> validateRequiredInputsArePresent(
+      List<PipelineInputDefinition> inputDefinitions, Map<String, Object> inputsMap) {
+    ArrayList<String> errorMessages = new ArrayList<>();
+    inputDefinitions.stream()
+        .filter(PipelineInputDefinition::getIsRequired)
+        .forEach(
+            inputDefinition -> {
+              String inputName = inputDefinition.getName();
+              if (!(inputsMap.containsKey(inputName))) {
+                errorMessages.add(String.format("%s is required", inputName));
+              }
+            });
+    return errorMessages;
+  }
+
+  /**
+   * Validate that no inputs are null
+   *
+   * @param inputDefinitions - list of input definitions for a pipeline
+   * @param inputsMap - map of inputs to validate
+   * @return list of error messages for inputs that are null
+   */
+  public List<String> validateNoNulls(
+      List<PipelineInputDefinition> inputDefinitions, Map<String, Object> inputsMap) {
+    ArrayList<String> errorMessages = new ArrayList<>();
+    inputDefinitions.forEach(
+        inputDefinition -> {
+          String inputName = inputDefinition.getName();
+          if (inputsMap.containsKey(inputName) && inputsMap.get(inputName) == null) {
+            errorMessages.add(String.format("%s must not be empty", inputName));
+          }
+        });
+    return errorMessages;
+  }
+
+  /**
+   * Validate that all present inputs are the correct type. We do not check for required inputs
+   * here, nor do we perform a check on null inputs.
+   *
+   * @param inputDefinitions - list of input definitions for a pipeline
+   * @param inputsMap - map of inputs to validate
+   * @return list of error messages for inputs that are not the correct type
+   */
+  public List<String> validateInputTypes(
+      List<PipelineInputDefinition> inputDefinitions, Map<String, Object> inputsMap) {
+    ArrayList<String> errorMessages = new ArrayList<>();
+    inputDefinitions.forEach(
+        inputDefinition -> {
+          String inputName = inputDefinition.getName();
+          // we assume we have already generated error messages for null inputs, so we can skip them
+          // here
+          if (inputsMap.containsKey(inputName) && inputsMap.get(inputName) != null) {
```

Comment:
alternatively we could just put the null validation in here, but this method was starting to feel heavy